tests/cancun/eip6780_selfdestruct/test_dynamic_create2_selfdestruct_collision.py

```diff
@@ -0,0 +1,158 @@
+
+
+@pytest.fixture
+def env():  # noqa: D103
+    return Environment(
+        coinbase="0x2adc25665018aa1fe0e6bc666dac8fc2697ff9ba",
+        difficulty=0x020000,
+        gas_limit=71794957647893862,
+        number=1,
+        timestamp=1000,
+    )
+
+
+@pytest.mark.valid_from("Merge")
+@pytest.mark.parametrize(
+    "create2_dest_already_in_state",
+    (False, True),
+)
+def test_dynamic_create2_selfdestruct_collision(
+    env: Environment,
+    fork: Fork,
+    create2_dest_already_in_state: bool,
+    state_test: StateTestFiller,
+):
+    """Dynamic Create2->Suicide->Create2 collision scenario:"""
+    """
+    Perform a CREATE2, make sure that the initcode sets at least a couple of storage keys,
+    then on a different call, in the same tx, perform a self-destruct.
+    Then:
+        a) on the same tx, attempt to recreate the contract   <=== Covered int this test
+            1) and create2 contract already in the state
+            2) and create2 contract is not in the state
+        b) on a different tx, attempt to recreate the contract
+    Perform a CREATE2, make sure that the initcode sets at least a couple of storage keys,
+    then in a different tx, perform a self-destruct.
+    Then:
+        a) on the same tx, attempt to recreate the contract
+        b) on a different tx, attempt to recreate the contract
+    Verify that the test case described
+    in https://wiki.hyperledger.org/pages/viewpage.action?pageId=117440824 is covered
+    """
+
+    first_create2_result = 1
+    second_create2_result = 2
+
+    address_to = to_address(0x0600)
+    address_code = to_address(0x0601)
+    address_create2_storage = to_address(0x0512)
+    suicide_destination = to_address(0x03E8)
+
+    create2_salt = 1
+    initcode = Initcode(
+        deploy_code=Op.SELFDESTRUCT(Op.PUSH20(suicide_destination)),
+        initcode_prefix=Op.SSTORE(1, 1)
+        + Op.CALL(Op.GAS(), Op.PUSH20(address_create2_storage), 0, 0, 0, 0, 0),
+    )
+    create2_address = compute_create2_address(address_code, create2_salt, initcode)
+
+    pre = {
+        address_to: Account(
+            balance=100000000,
+            nonce=0,
+            code=Op.CALLDATACOPY(0, 0, Op.CALLDATASIZE())
+            + Op.SSTORE(
+                1, Op.CALL(Op.GAS(), Op.PUSH20(address_code), 0, 0, Op.CALLDATASIZE(), 0, 0)
+            )
+            + Op.SSTORE(2, 1),
+            storage={0x01: 0xFF},
+        ),
+        address_code: Account(
+            balance=1000000000000000000,
+            nonce=0,
+            code=Op.CALLDATACOPY(0, 0, Op.CALLDATASIZE())
+            + Op.SSTORE(first_create2_result, Op.CREATE2(7, 0, Op.CALLDATASIZE(), create2_salt))
+            + Op.CALL(Op.GAS(), Op.SLOAD(1), 3, 0, 0, 0, 0)
+            + Op.CALLDATACOPY(0, 0, Op.CALLDATASIZE())
+            + Op.SSTORE(second_create2_result, Op.CREATE2(5, 0, Op.CALLDATASIZE(), create2_salt)),
+            storage={0x02: 0xFF},
```

Comment:
```suggestion
            storage={
                first_create2_result: 0xFF,
                second_create2_result: 0xFF,
            }
```
We need to also preset `0x01` to something non-zero, to make sure that the transaction did not OOG'd.

I tried this locally and at the moment it is still running out of gas if the contract was pre-existing, because the `CREATE2` fails and then consumes 63/64 of the remaining tx's gas, and it runs out of gas by the `Op.CALL(...`.

I still think we need to make a subcall to a contract that only does:
```python
Op.CALLDATACOPY(0, 0, Op.CALLDATASIZE())
+ Op.MSTORE(0, Op.CREATE2(7, 0, Op.CALLDATASIZE(), create2_salt))
+ Op.RETURN(0, 32)
```
And then call that contract with a fraction of the tx's total gas, so it only consumes that fraction with the failing `CREATE2` and it let's the top level contract finish.